cmd/stdiscosrv/database.go

```diff
@@ -0,0 +1,264 @@
+	inbox      chan func()
+	stop       chan struct{}
+	clock      clock
+	marshalBuf []byte
+}
+
+func newLevelDBStore(dir string) (*levelDBStore, error) {
+	db, err := leveldb.OpenFile(dir, levelDBOptions)
+	if err != nil {
+		return nil, err
+	}
+	return &levelDBStore{
+		db:    db,
+		inbox: make(chan func(), 16),
+		stop:  make(chan struct{}),
+		clock: defaultClock{},
+	}, nil
+}
+
+func (s *levelDBStore) put(key string, rec DatabaseRecord) error {
+	t0 := time.Now()
+	rc := make(chan error)
+	s.inbox <- func() {
+		if len(rec.Addresses) != 0 {
+			rec.Misses = 0
+			// grab the existing record
+			cur, err := s.get(key)
+			if err == nil {
+				// merge addresses if the get was successful
+				databaseMerges.Inc()
+				rec.Addresses = merge(cur.Addresses, rec.Addresses)
+			}
+		}
+		size := rec.Size()
+		if len(s.marshalBuf) < size {
+			s.marshalBuf = make([]byte, size)
+		}
+		n, _ := rec.MarshalTo(s.marshalBuf)
+		rc <- s.db.Put([]byte(key), s.marshalBuf[:n], nil)
+	}
+	databasePuts.Inc()
+	databasePutSeconds.Observe(time.Since(t0).Seconds())
+	return <-rc
+}
+
+func (s *levelDBStore) get(key string) (DatabaseRecord, error) {
+	keyBs := []byte(key)
+	val, err := s.db.Get(keyBs, nil)
+	databaseGets.Inc()
+	if err == leveldb.ErrNotFound {
+		return DatabaseRecord{}, nil
+	}
+	if err != nil {
+		return DatabaseRecord{}, err
+	}
+
+	var rec DatabaseRecord
+
+	if err := rec.Unmarshal(val); err != nil {
+		s.inbox <- func() {
+			s.db.Delete(keyBs, nil)
+			databaseDeletes.Inc()
+		}
+		return DatabaseRecord{}, nil
+	}
+
+	rec.Addresses = expire(rec.Addresses, s.clock.Now().UnixNano())
+	return rec, nil
+}
+
+func (s *levelDBStore) Serve() {
+	t := time.NewTimer(0)
+	defer t.Stop()
+	defer s.db.Close()
+
+	// Start the statistics serve routine. It will exit with us when
+	// statisticsTrigger is closed.
+	statisticsTrigger := make(chan struct{})
+	defer close(statisticsTrigger)
+	statisticsDone := make(chan struct{})
+	go s.statisticsServe(statisticsTrigger, statisticsDone)
+
+	for {
+		select {
+		case fn := <-s.inbox:
+			// Run function in serialized order.
+			fn()
+
+		case <-t.C:
+			// Trigger the statistics routine to do its thing in the
+			// background.
+			statisticsTrigger <- struct{}{}
+
+		case <-statisticsDone:
+			// The statistics routine is done with one iteratation, schedule
+			// the next.
+			t.Reset(databaseStatisticsInterval)
+
+		case <-s.stop:
+			// We're done.
+			return
+		}
+	}
+}
+
+func (s *levelDBStore) statisticsServe(trigger <-chan struct{}, done chan<- struct{}) {
+	for range trigger {
+		t0 := time.Now()
+		activeCutoff := t0.Add(-24 * time.Hour).UnixNano()
+		active, errors, negative := 0, 0, 0
+
+		iter := s.db.NewIterator(&util.Range{}, nil)
+		for iter.Next() {
+			// Attempt to unmarshal the record, delete it and count the
+			// failure if there's something wrong with it.
+			var rec DatabaseRecord
+			if err := rec.Unmarshal(iter.Value()); err != nil {
+				s.db.Delete(iter.Key(), nil)
```

Comment:
This is actually a "can't happen" condition, should probably be removed entirely rather than optimized. (It kicked in once when I changed serialization from JSON to proto...)